docs/rfcs/0561-list-with-metadata.md

```diff
@@ -0,0 +1,72 @@
+- Proposal Name: `list_with_metadata`
+- Start Date: 2022-08-23
+- RFC PR: None
+- Tracking Issue: None
+
+# Summary
+
+Add a `metadata` field for `DirEntry`s.
+
+```rust
+pub struct DirEntry {
+    acc: Arc<dyn Accessor>,
+
+    metadata: ObjectMetadata,
```

Comment:
I'm a bit worried about this.

We expect `ObjectMetadata` to contain the **full** metadata of this object which often returned by `stat`. On the other side, that means users couldn't fetch more metadata any more.

So, embed `ObjectMetadata` in `DirEntry` violate this expectation:

- `list` often don't have enough metadata like `stat`.
- `ObjectMetadata` has some assumetion like: `content_length()` returns `u64` instead of `Option<u64>`. But in `list`, some services may not have `size` returned, so we don't know which value to return.

Reply:
We can define another `Metadata` type like https://doc.rust-lang.org/std/fs/struct.Metadata.html#method.len

Reply:
Nice idea. But for now, I prefer to store inside `DirEntry` directly.